src/Draco.Compiler/Syntax/ParseTree.cs

```diff
@@ -0,0 +1,144 @@
+using Draco.Compiler.Utilities;
+
+namespace Draco.Compiler.Syntax;
+
+/// <summary>
+/// An immutable structure representing a parsed source text with information about concrete syntax.
+/// </summary>
+internal abstract record class ParseTree
+{
+    public sealed record class CompilationUnit(
+        ValueArray<Decl> Declarations) : ParseTree;
+
+    public abstract record class Decl : ParseTree
+    {
+        public sealed record class Func(
+            IToken FuncKeyword,
+            IToken Identifier,
+            IToken OpenParenToken,
+            ValueArray<(
+                (   IToken Identifier,
+                    TypeSpecifier Type
+                ) Param,
+                IToken CommaToken
+            )> Params,
+            IToken CloseParenToken,
+            TypeSpecifier? Type,
+            FuncBody Body) : Decl;
+
+        public sealed record class Label(
+            IToken Identifier,
+            IToken ColonToken) : Decl;
+
+        public sealed record class Variable(
+            IToken Keyword, // Either var or val
+            IToken Identifier,
+            TypeSpecifier? Type,
+            (   IToken EqualsToken,
+                Expr Expression
+            )? Initializer) : Decl;
+    }
+
+    public abstract record class FuncBody
+    {
+        public sealed record class BlockBody(
+            Block Block) : FuncBody;
+
+        public sealed record class InlineBody(
+            IToken EqualsToken,
+            Expr Expression) : FuncBody;
+    }
+
+    public abstract record class TypeName : ParseTree
```

Comment:
These should be called `TypeExpr` or something. Type name infers identifiers only, but types can be a whole set of expressions themselves.

Reply:
Was very unsure of what to call it. Thanks.